### ISSUE TITLE
Add URL.createObjectURL and URL.revokeObjectURL

### ISSUE THREAD

**Issue by fmatoss** [tmpvar/jsdom#1721]
Hi, `jsdom` already supports the `URL` constructor, but not these 2 static methods, they are supported in most modern browsers.

[createObjectURL](https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL)
[revokeObjectURL](https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL)
[Can I use](http://caniuse.com/#feat=bloburls)

**Comment by unional** [comment]
Can the implementation in https://github.com/eligrey/Blob.js be used in `jsdom`?

I'm blocked by this for using `jspdf` inside `jsdom`.

**Comment by domenic** [comment]
No, the code there appears to create data URLs, not blob URLs. Someone needs to properly implement the spec at https://w3c.github.io/FileAPI/#dfn-createObjectURL

**Comment by unional** [comment]
I'm not familiar with this, will it be something like this?

```js
createObjectURL(blob) {
  return `blob:${asciiSerialize(location.origin)}/${createUUID()}`;
}
```

If so, need to know which function `asciiSerialize()` and `createUUID()` to use.

**Comment by fmatoss** [comment]
@unional `createUUID` can be [uuidV4](https://github.com/kelektiv/node-uuid#uuidv4options--buffer--offset)

I think Chrome isn't following the spec strictly, `URL.revokeObjectURL` won't throw `NetworkError` if you pass the same args twice or any type not expected, like a number or array.

**Comment by unional** [comment]
That means:

```js
const uuid = require('uuid/4');

createObjectURL(blob) {
  var implementationDefinedValue = ???
  var url = `blob:${asciiSerialize(location.origin) || implementationDefinedValue}/${uuid()}`;
  return url;
}
revokeObjectURL(blobUrl) {
  return;
}

asciiSerialize(origin) {
  if (origin.scheme) {
    return `${origin.scheme}://${serializeHost(origin.host)}${origin.port? ':' + +origin.port}`;
  }
  else {
    return 'null';
  }
}

serializeHost(host) {
  ...
}
```

I guess there is a `serializeHost()` somewhere in the code already.

One last thing is:

> If serialized is "null", set it to an implementation-defined value.

What does "implementation-defined value" mean?

**Comment by unional** [comment]
Seems like I get all the questions answered.
http://stackoverflow.com/questions/42452543/what-does-implementation-defined-value-in-fileapi-unicodebloburl-mean/42452714#42452714

I think it is ready to be implemented on https://github.com/jsdom/whatwg-url/blob/master/lib/URL-impl.js
and update https://github.com/jsdom/whatwg-url/blob/master/src/URL.idl

**Comment by brettz9** [comment]
It ought to do more than just make the Blob URLs. It ought to subsequently allow a URL so generated to be provided to the likes of `XMLHttpRequest`.

**Comment by unional** [comment]
> It ought to do more than just make the Blob URLs. It ought to subsequently allow a URL so generated to be provided to the likes of XMLHttpRequest.

Does it need that? What is the use cases?  Would the existing logic of `XMLHttpRequest` handle it just well?

I'm not sure if the goal of `jsdom` is to replicate the exact behavior of what a browser does.
If that is the case, we are writing a browser.

IMO we should defer this until there is an actual use case.

What else can we do to move this forward?

**Comment by unional** [comment]
> var url = `blob:${serializeURL(location.origin)}/${uuid()}`;

The `serializeURL()` does a bit more than the algorithm described in the spec.
Should we use it or implement exactly as in the spec?

**Comment by domenic** [comment]
There's definitely no interest in adding object URLs to jsdom if they don't actually work (when used by XHR, img elements, etc.). Just creating them is very, very uninteresting. You can do that with a few lines of code; you don't need a whole jsdom implementation for that.

**Comment by unional** [comment]
Ok, so what's need to implement that?

**Comment by brettz9** [comment]
Besides allowing DOM APIs to be used in Node, it is very helpful to have such a browser for running automated tests.

Specifically, if you are testing browser/Node code where a Blob gets created, this method can be used in both environments to allow you to introspect on the contents (as the browser at least otherwise does not allow this). (In my case, it would allow me to avoid writing Node-specific code for cloning Blobs in the structured cloning algorithm as used by IndexedDB (and `postMessage`).)

Blob URLs meet a special need of working in memory independent of a server. I think there are three unique aspects of Blob URLs over `data:` URLs (besides being able to compose them more easily without escaping and such).

One is that beyond in-memory content , they can reference files without needing to encode the whole file's contents within the URL (and not requiring it to be dereferenced until used).

Two is that they might be useful for security/privacy in avoiding persistence of the data beyond a browser session or outside of the user's browser.

Third is that Blob URLs can be revoked even within a session, allowing references to expire (e.g., if you create an image, display it via a Blob URL, and then the user deletes the image, you can revoke the URL so that the Blob URL cannot be reused, unlike would be the case with say a `data` URL).

**Comment by stereobooster** [comment]
> There's definitely no interest in adding object URLs to jsdom if they don't actually work (when used by XHR, img elements, etc.). Just creating them is very, very uninteresting. 

I want to do server side rendering of React (CRA) project with the help of [react-snapshot](https://github.com/geelen/react-snapshot). My project also uses [react-mapbox-gl](https://github.com/alex3165/react-mapbox-gl), which in turn uses mapbox-gl, which uses [webworkify](https://github.com/substack/webworkify). And process fails because of 

```
Error: Uncaught [TypeError: o.URL.createObjectURL is not a function]
```

I'm not interested in actually rendering map on server, I just want to render placeholder instead of map on serve, but this error prevents it.

> You can do that with a few lines of code; you don't need a whole jsdom implementation for that.

Is there a way I can patch jsdom to at least add noop function for this? I suppose it will fail, but this at least will be starting point.

**Comment by domenic** [comment]
See the readme: https://github.com/tmpvar/jsdom#intervening-before-parsing

**Comment by stereobooster** [comment]
Thanks. It's a pity I'm [stuck](https://github.com/stereobooster/react-snapshot/blob/master/src/snapshot.js#L8) with old api e.g. `jsdom.env`, which doesn't have this

```js
options.beforeParse(this[window]._globalProxy);
```

(facepalm) it will be a bit harder

**Comment by fredvollmer** [comment]
What's the status of this feature request? It would certainly be useful to have.

**Comment by domenic** [comment]
https://mobile.twitter.com/slicknet/status/782274190451671040

**Comment by dylanjha** [comment]
@fredvollmer if using jest with jsdom and you want to noop this function you can add this code to `setupTest.js` (or an equivalent setup script)

```javascript
function noOp () { }

Object.defineProperty(window.URL, 'createObjectURL', { value: noOp})
```

**Comment by franciscolourenco** [comment]
@dylanjha I get `TypeError: Cannot redefine property: createObjectURL`

`jsdom@11.11.0`

**Comment by dylanjha** [comment]
@franciscolourenco you're right! Sorry about that... just updated the code in the above example to wrap in this conditional and that is working for me: `if (typeof window.URL.createObjectURL === 'undefined') {`

Give that a try!

Thanks for reminding me to come back here and update my example.

**Comment by franciscolourenco** [comment]
@dylanjha `createObjectURL` is defined but not a function, so some libraries throw errors. An assignment worked for me:

```
window.URL.createObjectURL = noOp
```

**Comment by PetterRuud** [comment]
Seems to be fixed with the version 1.7 for me.

**Comment by ryderdonahue** [comment]
Just to make it extra clear, doing something like 
```
function noOp () { }

if (typeof window.URL.createObjectURL === 'undefined') {
  Object.defineProperty(window.URL, 'createObjectURL', { value: noOp})
}
```
will ***only*** work if you put it within the "setupFiles" file of your jest configuration. So in your package.json or wherever you have
```"jest": {
"setupFiles": [
      "<rootDir>/internals/testing/setup_file.js"
    ],
}
```
you put that code written by others inside that file. (if that file doesnt exist you may need to create it)

Hope this helps anyone confused about the solution

**Comment by sumegha26** [comment]
img.src = URL.createObjectURL(blob);

not working, getting an error URL.createObjectURL is not a function while checking in jest test case file

**Comment by GMartigny** [comment]
This is hacky and I would not recommend it, but I made a simplified [implementation of `createObjectURL` and `revokeObjectURL` for Node.js](https://github.com/GMartigny/crop-browser/blob/master/test/_set-env.js).
Please make sure you know what you're doing before using it.

**Comment by peterpeterparker** [comment]
Late to the party but I solved it by adding the following in my `jest.setup.ts`:

```
window.URL.createObjectURL =
    window.URL.createObjectURL ||
    ((object: any) => {
        return null;
    });
```

**Comment by vdechef** [comment]
*(hope it can be usefull for someone ...)* I use Jest with FakeIndexedDB for my app's unit tests, and I was stucked because I needed URL.createObjectURL to work in jsdom. I could not use noOp, otherwise my tests would be meaningless.

I was finally able to implement a working mock for URL.createObjectURL, by using stuff from *eligrey/Blob.js*.  
See https://github.com/dumbmatter/fakeIndexedDB/issues/56#issuecomment-896353134

**Comment by lucaswiman** [comment]
[jsdom-worker](https://github.com/developit/jsdom-worker) adds a working implementation of `createObjectURL`, along with support for web workers.

**Comment by nemanjam** [comment]
> [jsdom-worker](https://github.com/developit/jsdom-worker) adds a working implementation of `createObjectURL`, along with support for web workers.

@lucaswiman do you have example code how to use it? I am trying to mock `Blob` and `URL.createObjectURL` but I get recursion. Here is my code:

```ts
import 'jsdom-worker';
import { Blob } from 'blob-polyfill';

global.Blob = Blob;

// TypeError: URL.createObjectURL is not a function
Object.defineProperty(window.URL, 'createObjectURL', {
  value: jest.fn().mockImplementation((file) => {
    let result = null;

    const code = `onmessage = e => postMessage(e.data)`;
    const worker = new Worker(URL.createObjectURL(new Blob([code])));
    worker.onmessage = (arg) => {
      result = arg;
    };
    worker.postMessage(file);

    return result;
  }),
});
```

**Comment by lucaswiman** [comment]
@nemanjam If you're using `jest`, then you need to add or update the jest configuration in package.json:
```js
  "jest": {
    "setupFiles": [
      "jsdom-worker"
    ]
  }
```

**Comment by nemanjam** [comment]
> @nemanjam If you're using `jest`, then you need to add or update the jest configuration in package.json:
> 
> ```js
>   "jest": {
>     "setupFiles": [
>       "jsdom-worker"
>     ]
>   }
> ```

It works, thanks, I lost 3 days to mock `Blob`, `File`and `createObjectURL` in jsdom.

**Comment by NameFILIP** [comment]
Good thread with a couple of workarounds, but it is still unclear: Is there a plan to implement `createObjectURL` inside of `jsdom`?

**Comment by domenic** [comment]
The plan is for you to send a pull request!

**Comment by geminiyellow** [comment]
still open from 2017, say hi from 2022.

```
TypeError: URL.revokeObjectURL is not a function
```

still here

**Comment by tkrotoff** [comment]
How to force Jest/jsdom to use Node.js `URL` implementation and thus `createObjectURL`, `revokeObjectURL`... (requires Node.js >= 16.7.0):

```TypeScript
// FixJSDOMEnvironment.ts

import JSDOMEnvironment from 'jest-environment-jsdom';

// https://github.com/facebook/jest/blob/v29.4.3/website/versioned_docs/version-29.4/Configuration.md#testenvironment-string
export default class FixJSDOMEnvironment extends JSDOMEnvironment {
  constructor(...args: ConstructorParameters<typeof JSDOMEnvironment>) {
    super(...args);

    // FIXME https://github.com/jsdom/jsdom/issues/1721
    this.global.URL = URL;
    this.global.Blob = Blob;
  }
}
```

```JavaScript
// jest.config.js

/** @type import('jest').Config */
const config = {
  testEnvironment: './FixJSDOMEnvironment.ts',

  ...
}

module.exports = config;
```

See also https://github.com/jsdom/jsdom/issues/1724#issuecomment-1446858041, https://github.com/jsdom/jsdom/issues/3363#issuecomment-1467894943

**Comment by csantos1113** [comment]
@tkrotoff thanks for your suggestion - I'm having an issue when trying to run that code:

```
● Validation Error:

  Test environment ./FixJSDOMEnvironment.js cannot be found. Make sure the testEnvironment configuration option points to an existing node module.

  Configuration Documentation:
  https://jestjs.io/docs/configuration
```

Did you run into the same issue?

**Comment by js2me** [comment]
I tried to switch vitest environment from `happy-dom` to `jsdom` and caused with various issues like this.  
Is this issue abandoned ?

**Comment by kirthi-b** [comment]
Want to take a crack at this, sticking to the File API spec: an origin-scoped blob URL store, `createObjectURL`/`revokeObjectURL` via webidl2js, and URL parsing that understands blob URLs. FileAPI/url WPTs as the acceptance suite.

First PR would skip fetch/XHR integration for blob URLs, that can be a follow-up. The store alone fixes the "createObjectURL is not a function" failure most test suites hit.

Scope look right?

**Comment by domenic** [comment]
No, incomplete spec non-compliant implementations are not accepted by the jsdom project.
